k2/python/k2/ctc_loss.py

```diff
@@ -0,0 +1,125 @@
+# Copyright (c)  2021  Xiaomi Corporation (authors: Fangjun Kuang)
+
+try:
+    from typing import Literal  # for Python >= 3.8
+except ImportError:
+    from typing_extensions import Literal  # for python < 3.8
+
+from typing import Optional
+
+import torch
+import torch.nn as nn
+
+from .autograd import intersect_dense
+from .dense_fsa_vec import DenseFsaVec
+from .fsa import Fsa
+
+
+class CtcLoss(nn.Module):
+    '''Ctc Loss computation in k2. It produces the same output as `torch.CtcLoss`
+    if given the same input.
+
+    One difference between `k2.CtcLoss` and `torch.CtcLoss` is that k2 accepts
+    a general FSA while PyTorch requires a linear FSA (represented as a list).
+    That means, `k2.CtcLoss` supports words with multiple pronunciations.
+
+    See `k2/python/tests/ctc_loss_test.py` for usage.
+
+    We assume that the blank label is always 0. The arguments `reduction` and
+    `target_lengths` have the same meaning as their counterparts in
+    `torch.CtcLoss`.
+    '''
+
+    def __init__(self):
+        super().__init__()
+
+    def forward(self,
+                decoding_graph: Fsa,
+                dense_fsa_vec: DenseFsaVec,
+                output_beam: float = 10,
+                reduction: Literal['none', 'mean', 'sum'] = 'sum',
+                use_double_scores: bool = True,
```

Comment:
I think that arguments `output_beam`, `reduction`, and `use_double_scores` could be moved into the constructor instead -- this would seem more consistent with `torch.CTCLoss` https://pytorch.org/docs/stable/generated/torch.nn.CTCLoss.html

Reply:
Thanks. Fixed.